src/main/scala/firrtl/passes/Checks.scala

```diff
@@ -155,47 +156,84 @@ object CheckHighForm extends Pass {
+        case t: WRef => literals.contains(t.name) && literals(t.name)
+        case t: WSubIndex => resolveToLiteral(literals, t.expr)
+        case t: WSubField => resolveToLiteral(literals, t.expr)
+        case t: WSubAccess => resolveToLiteral(literals, t.expr)
+        case _ => false
+      }
+    }
+    
+    def validateLiteral(info: Info, mname: String, literals: LiteralSet, loc: Expression, value: Expression): Unit = {
+      loc match {
+        case t: HasName => validateLiteralS(info, mname, literals, t.name, value)
+        case t: WRef => validateLiteralS(info, mname, literals, t.name, value)
+        case t: WSubIndex => validateLiteral(info, mname, literals, t.expr, value)
+        case t: WSubField => validateLiteral(info, mname, literals, t.expr, value)
+        case t: WSubAccess => validateLiteral(info, mname, literals, t.expr, value)
+        case _ => // do nothing 
+      }
+    }
+    
+    def validateLiteralS(info: Info, mname: String, literals: LiteralSet, name: String, value: Expression): Unit = {
+      if (literals.contains(name))
+        literals(name) = literals(name) && resolveToLiteral(literals, value)
+      else 
+        literals(name) = resolveToLiteral(literals, value)
+    }
 
-    def checkHighFormS(minfo: Info, mname: String, names: NameSet)(s: Statement): Unit = {
+    def checkHighFormS(minfo: Info, mname: String, names: NameSet, literals: LiteralSet)(s: Statement): Unit = {
       val info = get_info(s) match {case NoInfo => minfo case x => x}
       s foreach checkName(info, mname, names)
       s match {
         case DefRegister(info, name, tpe, _, reset, init) =>
+          literals(name) = false
           if (hasFlip(tpe))
             errors.append(new RegWithFlipException(info, mname, name))
-          if (reset.tpe == AsyncResetType && !init.isInstanceOf[Literal])
+          if (reset.tpe == AsyncResetType && !(resolveToLiteral(literals,init)))
             errors.append(new NonLiteralAsyncResetValueException(info, mname, name, init.serialize))
         case sx: DefMemory =>
+        literals(name) = false
           if (hasFlip(sx.dataType))
             errors.append(new MemWithFlipException(info, mname, sx.name))
           if (sx.depth <= 0)
             errors.append(new NegMemSizeException(info, mname))
         case sx: WDefInstance =>
+        literals(name) = false
           if (!moduleNames(sx.module))
             errors.append(new ModuleNotDefinedException(info, mname, sx.module))
           // Check to see if a recursive module instantiation has occured
           val childToParent = moduleGraph add (mname, sx.module)
           if (childToParent.nonEmpty)
             errors.append(new InstanceLoop(info, mname, childToParent mkString "->"))
         case sx: Connect => checkValidLoc(info, mname, sx.loc)
+                            validateLiteral(info, mname, literals, sx.loc, sx.expr)
+        case sx: DefWire => literals(sx.name) = true
```

Comment:
If I'm correct that this case is handling unconnected wires, rather than defaulting `DefWires` to literals, I would explicitly do this on `IsInvalid`, I'd rather the compiler crash if something goes wrong than hide a bug.

Reply:
Not exactly: this case is defining that any wire is a literal, whatever its connections/state/validy.
This is a potential flaw as I explained above (invalidating Connect might come after the DefRegister) however not the one your are talking about.
To confirm, I've just tried with the following test:

```
  "Unconnected Wire" should "NOT be allowed as reset values for AsyncReset" in {
    an [passes.CheckHighForm.NonLiteralAsyncResetValueException] shouldBe thrownBy {
      compileBody(s"""
        |input clock : Clock
        |input reset : AsyncReset
        |input x : UInt<1>
        |output z : UInt<1>
        |wire w : UInt<1>
        |reg r : UInt<1>, clock with : (reset => (reset, w))
        |r <= x
        |z <= r""".stripMargin
      )

```
and I got a `RefNotInitializedException` error from `firrtl.passes.CheckInitialization` as I would have expected (or at least hoped for ^^)
```
- should NOT be allowed as reset values for AsyncReset *** FAILED ***
[info]   Expected exception firrtl.passes.CheckHighForm$NonLiteralAsyncResetValueException to be thrown, but firrtl.passes.CheckInitialization$RefNotInitializedException was thrown (AsyncResetSpec.scala:133)
```